
#Data analysis process
* 1.Asking Question
* 2.Data Wrangling
* 3.Exploratory Data Analysis
* 4.Drawing Conclusion
* 5.Comunicating Results

In [ ]:
#2a.Gathering Data
#(csv files)(APIs)(Web Scraping)(Databases)

In [ ]:
#2b.Assessing Data
#1.Finding the number of rows/columns(shape)
#2.Data types of various columns(info())
#3.Checking for missing values(info())
#4.check for duplicate data (is_unique())
#5.memory occupied by the dataset(info)
#6.high level mathematical overview of the data(describe)

In [ ]:
#2c.Cleaning the data
#1.Missing Data(e.g. mean)
#2.Remove duplicate data(drop_duplicates)
#3.Incorrect data type(astype)

In [ ]:
#3a.Exploring data
#1.Finding Correlation and Covariance
#2.Doing univariate and multivariate analysis
#3.Plotting graphs(data visualization)


In [ ]:
#3b.Augmenting Data
#1.Removing Outliers
#2.Merging DataFrames
#3.Adding new Columns
#these operations are collectively called a feature engineering

In [ ]:
#4.drawing conclusion
#1.Machine learning
#2.Inferential Statistics
#3.Descriptive Statistics

In [ ]:
#5.Communicating Results/Data Storytelling
#In person
#Reports
#Blog Post
#PPTs/Slide decks
##comunication skill

In [ ]:
#you can do this steps in non linear manner
#there is no particular order to follow this steps

#Data Assessing

* Assessing data
- in this step, the data is to be understood more deeply.before implementing methods to clean it.you will definitely need to have a better idea about what the data ia about.

###Types of Unclean Data
-There are 2 kinds of unclean data
1. Dirty Data (Data with Quality issues):
Dirty data, also known as low quality data. Low quality data has content issues
2. Messy_ Data (Data with tidiness issues)
Messy data, also known as untidy data. Untidy data has structural issues.

-1. Dirty Data
When the data has following kinds of issues:
1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

-2. Messy Data
When the data has structural issue is it known as Messy or untidy data
Tidy data has the following properties
1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a fable
If a data violates any one of the above it is known as untidy data

- Types of Assessment
There are 2 types of assessment styles
1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

- Steps in Assessment
There are 2 steps involved in Assessment
1. Discover
2. Document



Data Quality Dimension
1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues

###Cleaning Data
you should follow the order provided below
- Completeness issuess
- tidiness issuess
- Remaining data quality issuess like validity,accuracy and consistency

Steps involved in data cleaning
- Define
- Code
- test

always make sure to create a copy of your pandas dataframe before you
start the cleaning process

In [ ]:
import pandas as pd


In [ ]:
patients=pd.read_csv('/content/drive/MyDrive/DATASETS/patients.csv')
treatments=pd.read_csv('/content/drive/MyDrive/DATASETS/treatments.csv')
adverse=pd.read_csv('/content/drive/MyDrive/DATASETS/adverse_reactions.csv')


In [ ]:
print(patients.shape)
print(treatments.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [ ]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [ ]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [ ]:
adverse

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

Quality
[Completeness,Validity,Accuracy,Consistency]
1. patients
- zip code is stored in decimal format[v]
- certain zip codes of 4 digits[v]
- for the state col sometimes full names are use and other times short forms are used[IC]
- missing values for contact details of 12 patients[C]
- errorneous data for weights and heights[?]
- inconsistent way of storing phone numbers in contact cols[A]
- duplicate and not useful data in dataset[IC]
2. treatments
- hba1c_change has missing value[C]
- novondra and auralin cols has u attached to numbers[V]
- only 280 rows are there , instead of 350[C]
- name and surname col is in lower case[IC]
- incorrect datatypes for a auralin and novadra cols[V]
- hbalc_change 9 instead of 4[A]
- use of dashes instead of Nan in aurolin and novodra cols[IC]
3. adverse
- name and surname col is in lowercase[IC]

Tidiness
1. patients
- phone and email together in one col i.e. contact
2. treatments
- 3 variables stored in auralin and novadra cols
3. adverse reaction
- no point of using this table

a. Completeness issues
- missing values contact details of 12 patients
- only 280 rows are there, instead of 350
- hbaic_change col has missing values

In [ ]:
# creating copies

patients_copy=patients.copy()
treatments_copy=treatments.copy()
adverse_copy=adverse.copy()

In [ ]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [ ]:
#subtract hbalc_start from hbalc_end to get hbalc_change
treatments_copy['hba1c_change']=treatments_copy['hba1c_start']-treatments_copy['hba1c_end']

In [ ]:
treatments_copy['hba1c_change'].isnull().sum()

0

In [ ]:
#novadra and aurol in using the melt function
pd.melt(treatment_copy, id_vars=['given_name','surname',''])